api/src/main/java/org/openmrs/module/addresshierarchy/service/AddressHierarchyServiceImpl.java

```diff
@@ -975,118 +1026,197 @@ private String encodeString(Method encodeStringMethod, String stringToEncode, St
 	}
 
-    @Override
+	@Override
 	@Transactional(readOnly = true)
-    public List<AddressHierarchyEntry> getAddressHierarchyEntriesByLevelAndLikeName(AddressHierarchyLevel level, String name, int limit) {
-        return dao.getAddressHierarchyEntriesByLevelAndLikeName(level, name, limit);    }
+	public List<AddressHierarchyEntry> getAddressHierarchyEntriesByLevelAndLikeName(AddressHierarchyLevel level, String searchString, int limit) {
+		if (level == null) {
+			return null;
+		}
+
+		List<AddressHierarchyEntry> matchedEntries = new ArrayList<AddressHierarchyEntry>();
+
+		if (!getI18nCache().isEnabled()) {
+			matchedEntries.addAll(dao.getAddressHierarchyEntriesByLevelAndLikeName(level, searchString, limit));
+		}
+		else {
+			for (String code : getI18nCache().getMessageKeysByLikeName(searchString)) {
+				matchedEntries.addAll(dao.getAddressHierarchyEntriesByLevelAndLikeName(level, code, limit));
+			}
+		}
+
+		return matchedEntries;
+	}
 
 	@Override
 	@Transactional(readOnly = true)
 	public AddressHierarchyEntry getAddressHierarchyEntryByUuid(String uuid) {
 		return dao.getAddressHierarchyEntryByUuid(uuid);
 	}
 
+	/*
+	 * I18N reverse cache convenience API
+	 */
+
+	protected AddressHierarchyI18nCache getI18nCache() {
+		if (ModuleFactory.isModuleStarted(AddressHierarchyConstants.MODULE_EXTI18N_ARTIFACT_ID)) {
+			return Context.getRegisteredComponent(ExtI18nConstants.COMPONENT_AH_REVI18N, AddressHierarchyI18nCache.class);
+		}
+		else {
+			return new EmptyI18nCache(); // an inert/disabled cache that does nothing
+		}
+	}
+
+	protected void initI18nCache(List<AddressHierarchyLevel> orderedLevels, AddressHierarchyI18nCache i18nCache, Locale locale) {
+		ListIterator<AddressHierarchyLevel> iterator = orderedLevels.listIterator(orderedLevels.size());
+		// Iterate in reverse.
+		while (iterator.hasPrevious()) {
+			for (AddressHierarchyEntry entry : getAddressHierarchyEntriesByLevel(iterator.previous())) {
+				i18nCache.getMessage(entry.getName(), locale); // this fills the cache down from the entry level up to the top
+			}
+		}
+	}
+
+	@Override
+	public void initI18nCache() {
+		AddressHierarchyI18nCache i18nCache = getI18nCache(); 
+		if (i18nCache.isEnabled()) {
+			List<AddressHierarchyLevel> orderedLevels = getOrderedAddressHierarchyLevels(false);
+			if (CollectionUtils.isEmpty(orderedLevels)) {
+				log.info("The Address Hierarchy reverse translation cache was not initialized because there are no mapped address hierarchy levels.");
+				return;
+			}
+			List<String> orderedAddressFields = new ArrayList<String>();
+			for (AddressHierarchyLevel level : orderedLevels) {
+				orderedAddressFields.add(level.getAddressField().getName());
+			}
+			i18nCache.setOrderedAddressFields(orderedAddressFields);
+			for (Locale locale : Context.getService(AdministrationService.class).getAllowedLocales()) {
+				initI18nCache(orderedLevels, i18nCache, locale);
+			}
+		}
+	}
+	
+	@Override
+	public void resetI18nCache() {

```

Comment:
Let's go ahead and move these service methods up higher in these file, before the utility methods and deprecated methods.